linux/tests/threadpool_linux_int/threadpool_linux_int.c

```diff
@@ -230,4 +276,302 @@ TEST_FUNCTION(threadpool_chaos_knight)
+}
+
+TEST_FUNCTION(stop_timer_waits_for_ongoing_execution)
+{
+    // assert
+    // create an execution engine
+    WAIT_WORK_CONTEXT wait_work_context;
+    volatile_atomic int32_t call_count;
+    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(NULL);
+    ASSERT_IS_NOT_NULL(execution_engine);
+
+    // create the threadpool
+    THREADPOOL_HANDLE threadpool = threadpool_create(execution_engine);
+    ASSERT_IS_NOT_NULL(threadpool);
+
+    // open
+    ASSERT_ARE_EQUAL(int, 0, threadpool_open_async(threadpool, on_threadpool_open_complete, NULL));
+
+    (void)interlocked_exchange(&wait_work_context.call_count, 0);
+    (void)interlocked_exchange(&wait_work_context.wait_event, 0);
+
+    // schedule one timer that waits
+    LogInfo("Starting timer");
+    TIMER_INSTANCE_HANDLE timer;
+    ASSERT_ARE_EQUAL(int, 0, threadpool_timer_start(threadpool, 0, 5000, wait_work_function, (void*)&wait_work_context, &timer));
+
+    // act
+    ThreadAPI_Sleep(500);
+
+    // call stop
+    LogInfo("Timer should be running and waiting, now stop timer");
+    threadpool_timer_destroy(timer);
+
+    LogInfo("Timer stopped");
+
+    // set the event, that would trigger a WAIT_OBJECT_0 if stop would not wait for all items
+    interlocked_increment(&wait_work_context.wait_event);
+
+    // cleanup
+    threadpool_close(threadpool);
+    threadpool_destroy(threadpool);
+    execution_engine_dec_ref(execution_engine);
+}
+
+TEST_FUNCTION(cancel_timer_waits_for_ongoing_execution)
+{
+    // assert
+    // create an execution engine
+    WAIT_WORK_CONTEXT wait_work_context;
+    volatile_atomic int32_t call_count;
+    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(NULL);
+    ASSERT_IS_NOT_NULL(execution_engine);
+
+    // create the threadpool
+    THREADPOOL_HANDLE threadpool = threadpool_create(execution_engine);
+    ASSERT_IS_NOT_NULL(threadpool);
+
+    // open
+    ASSERT_ARE_EQUAL(int, 0, threadpool_open_async(threadpool, on_threadpool_open_complete, NULL));
+
+    (void)interlocked_exchange(&wait_work_context.call_count, 0);
+    (void)interlocked_exchange(&wait_work_context.wait_event, 0);
+
+    // schedule one timer that waits
+    LogInfo("Starting timer");
+    TIMER_INSTANCE_HANDLE timer;
+    ASSERT_ARE_EQUAL(int, 0, threadpool_timer_start(threadpool, 0, 5000, wait_work_function, (void*)&wait_work_context, &timer));
+
+    // act
+    ThreadAPI_Sleep(500);
+
+    // call cancel
+    LogInfo("Timer should be running and waiting, now cancel timer");
+    threadpool_timer_cancel(timer);
+
+    LogInfo("Timer canceled");
+
+    // set the event, that would trigger a WAIT_OBJECT_0 if stop would not wait for all items
+    interlocked_increment(&wait_work_context.wait_event);
+
+    // cleanup
+    threadpool_timer_destroy(timer);
+    threadpool_close(threadpool);
+    threadpool_destroy(threadpool);
+    execution_engine_dec_ref(execution_engine);
+}
+
 END_TEST_SUITE(TEST_SUITE_NAME_FROM_CMAKE)
```

Comment:
>TEST_SUITE_NAME_FROM_CMAKE [](http://example.com/codeflow?start=15&length=26)

I am not going to stop progress with a wait.

"where specs?"

"where UTs"?

./signed off

Reply:
specs and UT will come soon.